File: DSA/Queue/q.go
package main

import "fmt"

type node struct {
	data int
	next *node
}
type que struct {
	head *node
	tail *node
}

func (q *que) addlast(e int) {
	new := &node{e, nil}
	if q.head == nil {
		q.head = new
		q.tail = new
	} else {
		q.tail.next = new
		q.tail = new

	}

}
func (q *que) removefirst() {
	if q.head == nil {
		fmt.Print("its empty")
		return
	} else {
		q.head = q.head.next
	}

}
func (q *que) addfirst(e int) {
	new := &node{e, nil}
	if q.head == nil {
		q.head = new
		q.tail = new
	} else {
		new.next = q.head
		q.head = new
	}

}

// func (q *que) removelast() {
// 	if q.head == nil {
// 		fmt.Print("its empty")
// 		return
// 	}
// 	p := q.head
// 	i := 0
// 	for p != nil {
// 		p = p.next
// 		i++
// 	}
// 	q.tail = p
// 	p = p.next
// 	q.tail.next = nil

// }
func (q *que) display() {
	p := q.head
	for p != nil {
		fmt.Print(p.data, "-->")
		p = p.next
	}

}

func main() {
	d := que{}
	d.addlast(10)
	d.addlast(20)
	d.addlast(30)
	d.display()
	fmt.Println()
	d.removefirst()
	d.display()
	fmt.Println()
	d.addfirst(40)
	d.display()
	// d.removelast()
	// d.display()

}

File: DSA/Queue/q.py
class node:
    def __init__(self,data,next):
        self.data=data
        self.next=None
class queue:
    def __init__(self):
        self.head=None
        self.tail=None
        self.size=0

    def __len__(self):
        return self.size

    def isempty(self):
        return self.size==0

    def enq(self,e):
        new=node(e,None)
        if self.isempty():
            self.head=new
            self.tail=new
        else:
            self.tail.next=new
            self.tail=new
        self.size+=1

    def deq(self):
        if self.isempty():
            print("its empty")
            return
        self.head=self.head.next
        self.size-=1


    
    def display(self):
        p=self.head
        while p!=None:
            print(p.data,end="-->")
            p=p.next
        print()
q=queue()
q.enq(1)
q.enq(2)
q.enq(3)
q.display()
q.deq()
q.display()